### PR TITLE
ENH: fix thread-unsafe C API usages

### PR DESCRIPTION
**Pull request by ngoldbaum**
Ref #26159 

See also the CPython HOWTO on this topic: https://docs.python.org/3.13/howto/free-threading-extensions.html#freethreading-extensions-howto.

The remaining usages of `PyDict_GetItem` and `PyDict_Next` are all around the `fields` attribute of structured dtypes. I'm pretty sure that dictionary is effectively frozen after the DType is constructed, so I don't worry about those uses.

It's not straightforward to write tests for this, I'm just applying static refactorings in places where the refactoring shouldn't introduce new reference counting bugs.

### CODE REVIEW

File: numpy/_core/src/multiarray/multiarraymodule.c [hunk 1]
```diff
@@ -2986,9 +2986,13 @@ array_einsum(PyObject *NPY_UNUSED(dummy), PyObject *args, PyObject *kwds)
     }
 
     /* Get the keyword arguments */
+    int error = 0;
     if (kwds != NULL) {
         PyObject *key, *value;
         Py_ssize_t pos = 0;
+#if Py_GIL_DISABLED
+        Py_BEGIN_CRITICAL_SECTION(kwds);
+#endif
```
Comment (seberg) [inline numpy/_core/src/multiarray/multiarraymodule.c:2995]: This is keyword parsing, you don't need it:
> keyword argument dictionaries in function calls; those keyword argument dictionaries are effectively private (not accessible by other threads), so using borrowed references in that context is safe.

Comment (seberg) [reply numpy/_core/src/multiarray/multiarraymodule.c:2995]: (Although we should just use `npy_parse_kwargs` or whatever it was here...  Will make this much faster and cleaner.)

**Review by seberg — COMMENTED** [review]
Second one seems unnnecessary to me, first one looks nice with `PyDict_GetItemRef` either way, I like it.
The last one seems good (I almost wonder if even a copy could make sense, but probably no reason to think about it).

File: numpy/_core/src/multiarray/multiarraymodule.c (hunk 1 above)
Comment (ngoldbaum) [reply numpy/_core/src/multiarray/multiarraymodule.c:2995]: Ah good point, thank you for helping me spot these forest-for-the-trees issues.

**Review by seberg — APPROVED** [review]
Thanks, now that I glance over, I half wonder if a `goto` solution can't be done nicely.  (I don't love the `error =` state)

But, I am not sure, and it looks good, so if you don't think of something quickly, please just merge.

**Comment by ngoldbaum** [comment]
The macro expands to statements with open and close braces so I couldn't think of a way to keep it syntactically sound and leave the gotos.

**Comment by seberg** [comment]
Ah, right, I hadn't realized that, thanks!

**Pull request merged**